🌟 Exercise 1 : Parsing HTML with BeautifulSoup
Instructions
Objective: Use urlopen() to fetch the HTML content of a webpage and then parse it using BeautifulSoup.



<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Sports World</title>
    <style>
        body { font-family: Arial, sans-serif; }
        header, nav, section, article, footer { margin: 20px; padding: 15px; }
        nav { background-color: #333; }
        nav a { color: white; padding: 14px 20px; text-decoration: none; display: inline-block; }
        nav a:hover { background-color: #ddd; color: black; }
        .video { text-align: center; margin: 20px 0; }
    </style>
</head>
<body>

    <header>
        <h1>Welcome to Sports World</h1>
        <p>Your one-stop destination for the latest sports news and videos.</p>
    </header>

    <nav>
        <a href="#football">Football</a>
        <a href="#basketball">Basketball</a>
        <a href="#tennis">Tennis</a>
    </nav>

    <section id="football">
        <h2>Football</h2>
        <article>
            <h3>Latest Football News</h3>
            <p>Read about the latest football matches and player news.</p>
            <div class="video">
                <iframe width="560" height="315" src="https://www.youtube.com/embed/football-video-id" frameborder="0" allowfullscreen>
                </iframe>
            </div>
        </article>
    </section>

    <section id="basketball">
        <h2>Basketball</h2>
        <article>
            <h3>NBA Highlights</h3>
            <p>Watch highlights from the latest NBA games.</p>
            <div class="video">
                <iframe width="560" height="315" src="https://www.youtube.com/embed/basketball-video-id" frameborder="0" allowfullscreen>
                </iframe>
            </div>
        </article>
    </section>

    <section id="tennis">
        <h2>Tennis</h2>
        <article>
            <h3>Grand Slam Updates</h3>
            <p>Get the latest updates from the world of Grand Slam tennis.</p>
            <div class="video">
                <iframe width="560" height="315" src="https://www.youtube.com/embed/tennis-video-id" frameborder="0" allowfullscreen></iframe>
            </div>
        </article>
    </section>

    <footer>
        <form action="mailto:contact@sportsworld.com" method="post" enctype="text/plain">
            <label for="name">Name:</label><br>
            <input type="text" id="name" name="name"><br>
            <label for="email">Email:</label><br>
            <input type="email" id="email" name="email"><br>
            <label for="message">Message:</label><br>
            <textarea id="message" name="message" rows="4" cols="50"></textarea><br><br>
            <input type="submit" value="Send">
        </form>
    </footer>

</body>
</html>


Read the HTML content of the page.
Create a BeautifulSoup object to parse this HTML.
Find the title of the webpage (the content inside the <title> tag).
Extract all paragraphs (<p> tags) from the page.
Retrieve all links (URLs in <a href=""> tags) on the page.


In [1]:
from bs4 import BeautifulSoup

In [2]:
html = """
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Sports World</title>
    <style>
        body { font-family: Arial, sans-serif; }
        header, nav, section, article, footer { margin: 20px; padding: 15px; }
        nav { background-color: #333; }
        nav a { color: white; padding: 14px 20px; text-decoration: none; display: inline-block; }
        nav a:hover { background-color: #ddd; color: black; }
        .video { text-align: center; margin: 20px 0; }
    </style>
</head>
<body>

    <header>
        <h1>Welcome to Sports World</h1>
        <p>Your one-stop destination for the latest sports news and videos.</p>
    </header>

    <nav>
        <a href="#football">Football</a>
        <a href="#basketball">Basketball</a>
        <a href="#tennis">Tennis</a>
    </nav>

    <section id="football">
        <h2>Football</h2>
        <article>
            <h3>Latest Football News</h3>
            <p>Read about the latest football matches and player news.</p>
            <div class="video">
                <iframe width="560" height="315" src="https://www.youtube.com/embed/football-video-id" frameborder="0" allowfullscreen>
                </iframe>
            </div>
        </article>
    </section>

    <section id="basketball">
        <h2>Basketball</h2>
        <article>
            <h3>NBA Highlights</h3>
            <p>Watch highlights from the latest NBA games.</p>
            <div class="video">
                <iframe width="560" height="315" src="https://www.youtube.com/embed/basketball-video-id" frameborder="0" allowfullscreen>
                </iframe>
            </div>
        </article>
    </section>

    <section id="tennis">
        <h2>Tennis</h2>
        <article>
            <h3>Grand Slam Updates</h3>
            <p>Get the latest updates from the world of Grand Slam tennis.</p>
            <div class="video">
                <iframe width="560" height="315" src="https://www.youtube.com/embed/tennis-video-id" frameborder="0" allowfullscreen></iframe>
            </div>
        </article>
    </section>

    <footer>
        <form action="mailto:contact@sportsworld.com" method="post" enctype="text/plain">
            <label for="name">Name:</label><br>
            <input type="text" id="name" name="name"><br>
            <label for="email">Email:</label><br>
            <input type="email" id="email" name="email"><br>
            <label for="message">Message:</label><br>
            <textarea id="message" name="message" rows="4" cols="50"></textarea><br><br>
            <input type="submit" value="Send">
        </form>
    </footer>

</body>
</html>
"""

In [3]:
soup = BeautifulSoup(html, "html.parser")

print(soup.prettify())

<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8"/>
  <meta content="width=device-width, initial-scale=1.0" name="viewport"/>
  <title>
   Sports World
  </title>
  <style>
   body { font-family: Arial, sans-serif; }
        header, nav, section, article, footer { margin: 20px; padding: 15px; }
        nav { background-color: #333; }
        nav a { color: white; padding: 14px 20px; text-decoration: none; display: inline-block; }
        nav a:hover { background-color: #ddd; color: black; }
        .video { text-align: center; margin: 20px 0; }
  </style>
 </head>
 <body>
  <header>
   <h1>
    Welcome to Sports World
   </h1>
   <p>
    Your one-stop destination for the latest sports news and videos.
   </p>
  </header>
  <nav>
   <a href="#football">
    Football
   </a>
   <a href="#basketball">
    Basketball
   </a>
   <a href="#tennis">
    Tennis
   </a>
  </nav>
  <section id="football">
   <h2>
    Football
   </h2>
   <article>
    <h3>
     Latest Football New

In [4]:
title = soup.title.get_text()

print(title)

Sports World


In [5]:
paragraphs = soup.find_all("p")

In [6]:
for paragraph in paragraphs:
    print(paragraph.get_text())

Your one-stop destination for the latest sports news and videos.
Read about the latest football matches and player news.
Watch highlights from the latest NBA games.
Get the latest updates from the world of Grand Slam tennis.


In [7]:
links = soup.find_all("a")

In [8]:
for link in links:
    print(link.get("href"))

#football
#basketball
#tennis


🌟 Exercise 2 : Scraping robots.txt from Wikipedia
Instructions
Write a Python program to download and display the content of robot.txt for wikipedia

In [9]:
from urllib.request import urlopen
from bs4 import BeautifulSoup
import requests

In [11]:
from urllib.request import Request, urlopen

url = "https://en.wikipedia.org/robots.txt"

headers = {
    "User-Agent": "BootcampPracticeBot/1.0 (https://github.com/auriane-bouaziz)"
}

request = Request(url, headers=headers)

with urlopen(request) as response:
    content = response.read().decode("utf-8")

print(content)

# robots.txt for http://www.wikipedia.org/ and friends
#
# Please note: There are a lot of pages on this site, and there are
# some misbehaved spiders out there that go _way_ too fast. If you're
# irresponsible, your access to the site may be blocked.
#

# Observed spamming large amounts of https://en.wikipedia.org/?curid=NNNNNN
# and ignoring 429 ratelimit responses, claims to respect robots:
# http://mj12bot.com/
User-agent: MJ12bot
Disallow: /

# advertising-related bots:
User-agent: Mediapartners-Google*
Disallow: /

# Wikipedia work bots:
User-agent: IsraBot
Disallow:

User-agent: Orthogaffe
Disallow:

# Crawlers that are kind enough to obey, but which we'd rather not have
# unless they're feeding search engines.
User-agent: UbiCrawler
Disallow: /

User-agent: DOC
Disallow: /

User-agent: Zao
Disallow: /

# Some bots are known to be trouble, particularly those designed to copy
# entire sites. Please obey robots.txt.
User-agent: sitecheck.internetseer.com
Disallow: /

User-agent: 

🌟 Exercise 3 : Extracting Headers from Wikipedia’s Main Page
Instructions
Write a Python program to extract and display all the header tags from wikipedia.



In [12]:
import requests
from bs4 import BeautifulSoup

url = "https://en.wikipedia.org/"

response = requests.get(url)
print(response.status_code)

403


In [13]:
from google.colab import files
from bs4 import BeautifulSoup

uploaded = files.upload()

html = next(iter(uploaded.values()))
soup = BeautifulSoup(html, "html.parser")

Saving Wikipedia, the free encyclopedia.html to Wikipedia, the free encyclopedia.html


In [14]:
headers = soup.find_all(["h1", "h2", "h3", "h4", "h5", "h6"])

for header in headers:
    print(header.name, ":", header.get_text(strip=True))

h1 : Main Page
h1 : Welcome toWikipedia
h2 : From today's featured article
h2 : Did you know...
h2 : In the news
h2 : On this day
h2 : Today's featured picture
h2 : Other areas of Wikipedia
h2 : Wikipedia's sister projects
h2 : Wikipedia languages


🌟 Exercise 4 : Checking for Page Title
Instructions
Write a Python program to check whether a page contains a title or not.



In [15]:
# Exercise 4: Checking for Page Title

if soup.title is not None:
    print("Page title:", soup.title.get_text(strip=True))
else:
    print("This page has no title.")

Page title: Wikipedia, the free encyclopedia


🌟 Exercise 5 : Analyzing US-CERT Security Alerts
Instructions
Write a Python program to get the number of security alerts issued by US-CERT in the current year.
Source

In [16]:
import requests
from bs4 import BeautifulSoup
from datetime import datetime

# Define the target URL
url = "https://www.cisa.gov/news-events/cybersecurity-advisories"

# Define filters
params = {
    "f[0]": "advisory_type:93",
    "f[1]": f"release_date_year:{datetime.now().year}"
}

# Fetch the webpage
response = requests.get(url, params=params, timeout=15)

print("Status code:", response.status_code)
print("URL:", response.url)

Status code: 200
URL: https://www.cisa.gov/news-events/cybersecurity-advisories?f%5B0%5D=advisory_type%3A93&f%5B1%5D=release_date_year%3A2026


In [17]:
import re

# Parse HTML
soup = BeautifulSoup(response.text, "html.parser")

# Extract page text
text = soup.get_text(" ", strip=True)

# Find the total number of results
match = re.search(
    r"Showing\s+[\d,]+\s+to\s+[\d,]+\s+of\s+([\d,]+)\s+results",
    text,
    re.IGNORECASE
)

if match:
    total_alerts = int(match.group(1).replace(",", ""))
    print(f"Total security alerts: {total_alerts}")
else:
    print("Result counter not found. Check the page structure and filters.")

Total security alerts: 5078


### Exercise 5: Analyzing US-CERT Security Alerts

**Objective:** Retrieve and count the security alerts published by CISA during the current year (2026).

**Method:**

* Used `requests` to retrieve the CISA cybersecurity advisories webpage.
* Applied URL parameters to filter publications by alert type and year.
* Used BeautifulSoup and regular expressions to extract the number of results.
* Checked whether the requested filters were correctly applied.

**Result:**

The script successfully retrieved the webpage and extracted a total of 5,078 results. However, further validation showed that CISA had not applied the requested filters.

Additional attempts to access the filtered page were unsuccessful due to website access restrictions.

Therefore, the exact number of security alerts published in 2026 could not be reliably determined.


🌟 Exercise 6 : Scraping Movie Details
Instructions
Write a Python program to get movie name, year and a brief summary of the top 10 random movies from this IMBD website.

# Exercise 6: Scraping Movie Details

## Objective

Extract the movie title, release year, and a brief summary
for 10 movies from an IMDb list using BeautifulSoup.

## Methodology

The original IMDb list provided in the exercise was private
and could not be accessed.

I therefore selected an alternative public IMDb list
containing 10 movies released in 2020.

Since IMDb restricted automated access from Google Colab,
I downloaded the HTML page manually and imported it into
the notebook.

Using BeautifulSoup, I inspected the HTML structure,
identified the relevant movie containers, and extracted
the required information.

Finally, I organized the results into a Pandas DataFrame
and checked for missing values.

Note: The alternative list contains 10 predefined movies,
rather than a random selection from the original list.

In [35]:
# Import required libraries
from google.colab import files
from bs4 import BeautifulSoup
import pandas as pd
import re


In [22]:
from google.colab import files

uploaded = files.upload()

html = next(iter(uploaded.values()))
soup = BeautifulSoup(html, "html.parser")

Saving Top 10 Movies of 2020.html to Top 10 Movies of 2020.html


In [36]:
# Exercise 6: Extract movie details

# Select movie title links
movie_titles = soup.select('a[href*="ls_t_"]')

movies_data = []

for movie_link in movie_titles:

    # 1. Extract movie title
    title = re.sub(
        r"^\d+\.\s*",
        "",
        movie_link.get_text(strip=True)
    )

    # 2. Extract release year
    metadata = list(movie_link.parents)[3].get_text(" ", strip=True)

    year_match = re.search(r"\b(?:19|20)\d{2}\b", metadata)

    year = int(year_match.group()) if year_match else None

    # 3. Find the movie container
    movie = movie_link.find_parent(
        "li",
        class_="ipc-metadata-list-summary-item"
    )

    # 4. Extract movie summary
    summary_element = movie.select_one(
        ".ipc-html-content-inner-div"
    )

    summary = (
        summary_element.get_text(" ", strip=True)
        if summary_element else None
    )

    # 5. Store extracted information
    movies_data.append({
        "Movie": title,
        "Year": year,
        "Summary": summary
    })

# Create the final DataFrame
df_movies = pd.DataFrame(movies_data)

display(df_movies)

,Movie,Year,Summary
0,Tenet,2020,Un espion est sélectionné pour déjouer un comp...
1,The Banker,2020,Années 60 : deux entrepreneurs Afro-Américains...
2,Les Sept de Chicago,2020,L'histoire de 7 personnes jugées à la suite de...
3,Charm City Kings,2020,"Mouse, quatorze ans, veut désespérément rejoin..."
4,Minari,2020,Une famille coréenne déménage en Arkansas pour...
5,La Mission,2020,Un vétéran de la guerre civile accepte de rame...
6,Soul,2020,Un musicien qui a perdu sa passion pour la mus...
7,USS Greyhound : La Bataille de l'Atlantique,2020,"Au début de la Seconde Guerre mondiale, un cap..."
8,One Night in Miami,2020,Un récit fictif d'une nuit incroyable où les i...
9,Underwater,2020,Une équipe de chercheurs aquatiques essayent d...


In [33]:
print("Number of movies:", len(df_movies))

print("\nMissing values:")
print(df_movies.isna().sum())

Number of movies: 10

Missing values:
Movie      0
Year       0
Summary    0
dtype: int64


## Results and Conclusion

Successfully extracted 10 movies from an alternative
IMDb list using BeautifulSoup.

The final DataFrame contains:
- Movie title
- Release year
- Movie summary

Data validation confirmed:
- 10 movies extracted
- No missing values
- All extracted movies were released in 2020

This exercise demonstrated how to inspect HTML structures,
identify relevant elements using CSS selectors, extract
information, and organize unstructured web data into
a structured Pandas DataFrame.

Limitations:
The HTML file was downloaded manually due to website
access restrictions. Therefore, the data extraction was
automated, but the webpage retrieval was not.